lib/Mojo/Redis2/Cursor.pm

```
@@ -0,0 +1,275 @@
+package Mojo::Redis2::Cursor;
+use Mojo::Base '-base';
+
+use Carp 'croak';
+use Scalar::Util 'looks_like_number';
+
+has 'redis';
+has _args => sub { [] };
+has '_command';
+has _cursor => 0;
+
+sub again {
+  my $self = shift;
+  $self->_cursor(0);
+  delete $self->{_finished};
+  return $self;
+}
+
+sub command { shift->{_command}[0] }
+
+sub finished { !!shift->{_finished} }
+
+sub hgetall {
+  my $cur = shift->_clone('HSCAN', shift);
+  return $cur->slurp(@_);
+}
+
+sub hkeys {
+  my $cb = ref $_[-1] eq 'CODE' ? pop : undef;
+  my $cur = shift->_clone('HSCAN' => @_);
+  my $wrapper = sub {
+    my $keys = [grep { $a = !$a } @{$_[2] || []}];
+    return $cur->$cb($_[1], $keys);
+  };
+  my $resp = $cur->slurp($cb ? ($wrapper) : ());
+  return $resp if $cb;
+  $cb = sub { $_[2] };
+  return $wrapper->(undef, '', $resp);
+}
+
+sub key { shift->{_command}[1] }
+
+sub keys {
+  my $cur = shift->_clone('SCAN');
+  unshift @_, 'MATCH' if $_[0] && !ref $_[0];
+  return $cur->slurp(@_);
+}
+
+sub new {
+  my $self    = shift->SUPER::new();
+  my $command = shift;
+  $self->_command($command eq 'SCAN' ? [$command] : [$command, shift]);
+  croak 'ERR Should not specify cursor value manualy.'
+    if looks_like_number $_[0];
```

No need for input check. The command should just fail in the redis if invalid input is given.